pkg/dogstatsd/server.go

```diff
@@ -260,77 +251,107 @@ func nextMessage(packet *[]byte) (message []byte) {
 	return message
 }
 
-func (s *Server) parsePacket(packet *listeners.Packet, metricSamples []*metrics.MetricSample, events []*metrics.Event, serviceChecks []*metrics.ServiceCheck) ([]*metrics.MetricSample, []*metrics.Event, []*metrics.ServiceCheck) {
-	extraTags := s.extraTags
+func (s *Server) parsePackets(batcher *batcher, packets []*listeners.Packet) {
+	for _, packet := range packets {
+		originTags := findOriginTags(packet.Origin)
+		//log.Tracef("Dogstatsd receive: %s", packet.Contents)
+		for {
+			message := nextMessage(&packet.Contents)
+			if message == nil {
+				break
+			}
+			if s.Statistics != nil {
+				s.Statistics.StatEvent(1)
+			}
+			messageType := findMessageType(message)
 
-	log.Tracef("Dogstatsd receive: %s", packet.Contents)
-	if packet.Origin != listeners.NoOrigin {
-		originTags, err := tagger.Tag(packet.Origin, tagger.DogstatsdCardinality)
-		if err != nil {
-			log.Errorf(err.Error())
-		} else {
-			extraTags = append(extraTags, originTags...)
+			switch messageType {
+			case serviceCheckType:
+				serviceCheck, err := s.parseServiceCheckMessage(message)
+				if err != nil {
+					log.Errorf("Dogstatsd: error parsing service check: %s", err)
+					continue
+				}
+				serviceCheck.Tags = append(serviceCheck.Tags, originTags...)
+				batcher.appendServiceCheck(serviceCheck)
+			case eventType:
+				event, err := s.parseEventMessage(message)
+				if err != nil {
+					log.Errorf("Dogstatsd: error parsing event: %s", err)
+					continue
+				}
+				event.Tags = append(event.Tags, originTags...)
+				batcher.appendEvent(event)
+			case metricSampleType:
+				sample, err := s.parseMetricMessage(message)
+				if err != nil {
+					log.Errorf("Dogstatsd: error parsing metrics: %s", err)
+					continue
+				}
+				if s.debugMetricsStats {
+					s.storeMetricStats(sample.Name)
+				}
+				sample.Tags = append(sample.Tags, originTags...)
+				batcher.appendSample(sample)
+				if s.histToDist && sample.Mtype == metrics.HistogramType {
```

Comment:
Appending a second copy if it is an `Histogram` is the wanted behavior? (2 samples are added to the batch here)

Reply:
Yup, see the original logic https://github.com/DataDog/datadog-agent/blob/d2bba60b17c0020f2894ee46aa0df6222af7b16a/pkg/dogstatsd/server.go#L325-L329